Add native masked_softmax

zrphercule opened the pull request:
Summary:
This diff enabled native masked softmax on CUDA, also expanded our current warp_softmax to accept masking.
The mask in this masked softmax has to be the same shape as input, and has to be contiguous.

In a following diff I will submit later, I will have encoder mask layout included, where input is BHDD and mask is BD.

Test Plan: buck build mode/opt -c fbcode.enable_gpu_sections=true caffe2/test:nn && buck-out/gen/caffe2/test/nn\#binary.par -r test_masked_softmax

Differential Revision: D32338419

vadimkantorov commented:
Would also be nice to have masked_max etc. May be useful for it to accept a "fill value" for the case when the mask is empty for the whole reduction domain

zrphercule commented:
> 

Sure, we can do that after. 
This masked_softmax is mainly used for Transformer, we would like to increase our transformer encoder's performance via this.

zrphercule2 committed:
Add native masked_softmax (#69268)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/69268

This diff enabled native masked softmax on CUDA, also expanded our current warp_softmax to accept masking.
The mask in this masked softmax has to be the same shape as input, and has to be contiguous.

In a following diff I will submit later, I will have encoder mask layout included, where input is BHDD and mask is BD.

Test Plan: buck build mode/opt -c fbcode.enable_gpu_sections=true caffe2/test:nn && buck-out/gen/caffe2/test/nn\#binary.par -r test_masked_softmax

Reviewed By: ngimel

Differential Revision: D32338419

fbshipit-source-id: f4a57792624e9cfe4986d4324deb24e0130c5608
